serialization.rst

```diff
@@ -0,0 +1,337 @@
+
+Examples
+~~~~~~~~
+
+::
+
+  stdClass {
+    public $foo => 42
+  } => { 'foo' : 42 }
+
+  MyClass {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+  } => { 'foo' : 42 }
+
+  AnotherClass implements Serializable {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+    function bsonSerialize() : array {
+        return [ 'foo' => $this->foo, 'prot' => $this->prot ];
+    }
+  } => { 'foo' : 42, 'prot' : "wine" }
+
+  UpperClass implements Persistable {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+    function bsonSerialize() : array {
+        return [ 'foo' => $this->foo, 'prot' => $this->prot ];
+    }
+  } => { 'foo' : 42, 'prot' : "wine", '__pclass' : Binary(0x80, "UpperClass") }
+
+
+Deserialization from BSON
+=========================
+
+For compound types, there are three data types:
+
+- ``root``: refers to the top-level BSON document *only*
+- ``document``: refers to nested BSON document *only*
+- ``array``: refers to a BSON array
+
+Each of those three data types can be mapped against different PHP types. The
+possible mapping values are:
+
+- *not set* or ``NULL`` — this is the default.
+
+  - A BSON array will be deserialized as a PHP ``array``.
+  - A BSON document (root or nested) without ``__pclass`` property [1]_ becomes a
+    PHP ``stdClass`` object, with each BSON document property becoming a
+    public ``stdClass`` property.
+  - A BSON document (root or nested) with ``__pclass`` property [1]_ becomes
+    a PHP object of the class name as defined by the ``__pclass`` property.
+
+    If the named class implements the ``MongoDB\BSON\Unserializable``
+    interface, then the properties of the BSON document, minus the
+    ``__pclass`` property are sent as an associative array to the
+    ``bsonUnserialize`` function to initialise the object's properties.
+    
+    If the named class does not implement the ``MongoDB\BSON\Unserializable``
+    interface, then the object gets initialized with the properties
+    **without** the ``__pclass`` property being one of them. It simply gets
+    left out.
+
+    The rationale for the latter is because no explicit choice about whether
+    to use ODM features has been made, it should not bail out because of an
+    extra ``__pclass`` property. The behaviour is different if the data type
+    was mapped to ``"odm"`` (see below).
+
+- ``"array"`` — turns a BSON array or BSON document into a PHP array.
+  ``__pclass`` properties [1]_ are igored.
+
+- ``"object"`` or ``"stdClass"`` — turns a BSON array or BSON document into a
+  ``stdClass`` object. There will be no special treatment of a ``__pclass``
+  property [1]_, buy it should **not** be set as property in the returned
+  object.
+
+- ``"odm"`` — uses the class name set in the ``__pclass`` property of a BSON
+  array or BSON document to determine the class name of the created object.
```

Comment:
> What if I want all embedded documents to be 'array', except for __pclass? How do I express that?

Is that why you originally implemented the ODM auto-detection (i.e. reading `__pclass` binary data) as taking precedence over every type map option? My issue with that was allowing a user to actually get `array` or `stdClass` back despite the BSON data containing `__pclass`, which is why I proposed that `__pclass` only supersede `PHONGO_TYPEMAP_NONE` in [PHPC-329](https://jira.mongodb.org/browse/PHPC-329).

If you actually have very different needs for embedded documents, perhaps [PHPC-314](https://jira.mongodb.org/browse/PHPC-314) is the solution. With that, I would imagine the "document" and "array" type map keys would only apply if we didn't have a more specific mapping that pertained to a field path.